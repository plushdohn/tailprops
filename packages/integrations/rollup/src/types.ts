import { AstroIntegration } from "./handlers/astro";

type FrameworkOptions =
  | {
      framework: "svelte-ssr" | "react";
    }
  | {
      framework: "astro";
      integrations?: AstroIntegration[];
    }
  | {
      framework: "preact";
      classAttribute?: string;
    };

export type TailpropsPluginOptions = FrameworkOptions & {
  debug?: boolean;
};
