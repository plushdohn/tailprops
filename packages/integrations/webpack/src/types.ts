export type TailpropsPluginOptions = {
  framework: "react" | "preact";
} & {
  dry?: boolean;
  keep?: string | RegExp | ((filename: string) => boolean);
};
