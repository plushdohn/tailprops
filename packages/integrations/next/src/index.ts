import type { NextConfig } from "next";
import {
  TailpropsPluginOptions,
  TailpropsWebpackPlugin,
} from "tailprops-webpack-plugin";

export function withTailprops(
  nextConfig: NextConfig,
  options: TailpropsPluginOptions = { framework: "react" }
) {
  const modifiedConfig: NextConfig = {
    ...nextConfig,
    webpack(config: any, context: any) {
      config.plugins.unshift(new TailpropsWebpackPlugin(options));

      if (nextConfig.webpack && typeof nextConfig.webpack === "function") {
        return nextConfig.webpack(config, context);
      }

      return config;
    },
  };

  return modifiedConfig;
}
