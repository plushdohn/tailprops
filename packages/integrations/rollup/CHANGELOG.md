# rollup-plugin-tailprops

## 0.2.0

### Minor Changes

- 6af425e: Removed support for curly braces

### Patch Changes

- Updated dependencies [6af425e]
  - tailprops@0.2.0

## 0.1.0

### Minor Changes

- 46a3b1c: Initial release

### Patch Changes

- f81004c: Added READMEs
- Updated dependencies [46a3b1c]
- Updated dependencies [f81004c]
  - tailprops@0.1.0
