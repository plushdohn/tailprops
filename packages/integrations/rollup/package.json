{
  "name": "rollup-plugin-tailprops",
  "version": "0.2.0",
  "description": "Rollup plugin for Tailprops",
  "main": "./dist/cjs/index.js",
  "module": "./dist/esm/index.js",
  "types": "./dist/esm/index.d.ts",
  "files": [
    "dist"
  ],
  "license": "MIT",
  "private": false,
  "publishConfig": {
    "access": "public"
  },
  "scripts": {
    "build": "pnpm build:commonjs && pnpm build:esm",
    "build:commonjs": "rimraf dist/cjs && tsc --module commonjs --outDir dist/cjs",
    "build:esm": "rimraf dist/esm && tsc --module esnext --outDir dist/esm",
    "test": "jest",
    "coverage": "jest --coverage"
  },
  "dependencies": {
    "tailprops": "workspace:^0.2.0"
  },
  "peerDependencies": {
    "rollup": "^3.15.0"
  },
  "devDependencies": {
    "@types/jest": "^29.4.0",
    "@types/node": "^18.13.0",
    "jest": "^29.4.2",
    "rollup": "^3.15.0",
    "ts-jest": "^29.0.5",
    "typescript": "^4.9.5"
  }
}
