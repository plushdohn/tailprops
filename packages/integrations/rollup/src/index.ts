import type { SourceDescription, TransformHook } from "rollup";
import { handleAstro } from "./handlers/astro";
import { handlePreact } from "./handlers/preact";
import { handleReact } from "./handlers/react";
import { handleSvelteSsr } from "./handlers/svelteSsr";
import { TailpropsPluginOptions } from "./types";

export function tailpropsPlugin(options: TailpropsPluginOptions) {
  const transform: TransformHook = function (code, id) {
    const context = this.meta;

    let out: Partial<SourceDescription> | null | undefined;

    if (options.framework === "svelte-ssr") {
      out = handleSvelteSsr({ code, id }, { ssr: context.watchMode });
    } else if (options.framework === "astro") {
      out = handleAstro({ code, id }, { integrations: options.integrations });
    } else if (options.framework === "react") {
      out = handleReact({ code, id });
    } else if (options.framework === "preact") {
      out = handlePreact(
        { code, id },
        { classAttribute: options.classAttribute }
      );
    }

    if (out === undefined) {
      throw new Error(
        `Framework not supported by the Tailprops Rollup plugin. If you think it's horribly bad that we don't support this framework, please open an issue on GitHub.`
      );
    }

    if (out === null) return null;

    return {
      ...out,
      map: this.getCombinedSourcemap() ? out.map : null,
      ast: undefined, // Can't reuse babel ASTs
    };
  };

  return {
    name: "tailprops",
    transform,
  };
}
