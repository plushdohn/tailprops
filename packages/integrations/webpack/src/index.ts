import { CleanPlugin, Compiler } from "webpack";
import { TailpropsPluginOptions as PluginOptions } from "./types";

export class TailpropsWebpackPlugin extends CleanPlugin {
  pluginOptions: PluginOptions;

  constructor(options: PluginOptions) {
    super(options);
    this.pluginOptions = options;
  }

  apply(compiler: Compiler): void {
    compiler.options.module.rules.unshift({
      test: /\.(tsx|ts|js|jsx)$/,
      exclude: /node_modules/,
      use: [
        {
          loader: require.resolve("./loader"),
          options: this.pluginOptions,
        },
      ],
    });
  }
}

export type TailpropsPluginOptions = PluginOptions;
