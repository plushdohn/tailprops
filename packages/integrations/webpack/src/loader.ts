import type { LoaderContext } from "webpack";
import { transpileJsUsingPropsObject } from "tailprops";
import { TailpropsPluginOptions } from "./types";

export default async function (
  this: LoaderContext<TailpropsPluginOptions>,
  input: string,
  sourceMap?: any
) {
  this.cacheable(true);

  const options = this.getOptions();

  const webpackCallback = this.async();

  if (options.framework === "react" || options.framework === "preact") {
    let out: ReturnType<typeof transpileJsUsingPropsObject>;
    try {
      out = transpileJsUsingPropsObject(input, {
        classAttribute: "className",
        generateSourceMaps: sourceMap,
      });
    } catch (err) {
      return webpackCallback(new Error(`${(err as Error).message}`));
    }

    return webpackCallback(null, out.code, out.map ?? undefined);
  }

  return webpackCallback(
    new Error(
      `Framework '${options.framework}' is not supported by Tailprops. If you think it's horribly bad that we don't support this framework, please open an issue on GitHub.`
    )
  );
}
