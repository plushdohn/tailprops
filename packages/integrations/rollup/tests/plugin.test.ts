import { SourceDescription } from "rollup";
import { tailpropsPlugin } from "../src/index";

const mockedTransformContext = {
  meta: {
    watchMode: true,
  },
  getCombinedSourcemap: () => null,
} as any;

describe("rollup plugin", () => {
  let plugin = tailpropsPlugin({ framework: "svelte-ssr" });

  it("should throw on unknown frameworks", () => {
    const instance = tailpropsPlugin({
      framework: "somethingthatdoesntexist" as any,
    });

    expect(() =>
      instance.transform.bind(mockedTransformContext)(
        "console.log('test')",
        "1283901.unknown"
      )
    ).toThrow();
  });

  describe("svelte integration", () => {
    beforeEach(() => {
      plugin = tailpropsPlugin({ framework: "svelte-ssr" });
    });

    it("should not touch non-svelte files", () => {
      const result = plugin.transform.bind(mockedTransformContext)(
        "console.log('test')",
        "1283901.unknown"
      );

      expect(result).toBeNull();
    });

    it("should parse but not alter files without tailprops", () => {
      const source = 'console.log("test");';

      const result = plugin.transform.bind(mockedTransformContext)(
        source,
        "1283901.svelte"
      ) as Partial<SourceDescription>;

      expect(result.code).toBe(source);
    });

    it("should apply template literals transform on SSR files", () => {
      const source =
        'create_ssr_component(`<p class="${"some-class"}" tw="${"bg-red-500"}" tw-2xl="${"bg-red-300"}">test</p>`);';

      const instance = tailpropsPlugin({ framework: "svelte-ssr" });

      const result = instance.transform.bind(mockedTransformContext)(
        source,
        "test.svelte"
      ) as Partial<SourceDescription>;

      expect(result.code).toMatchSnapshot();
    });

    it("should change attribute function on build vs preview", () => {
      const source =
        'create_ssr_component(`<p class="${"some-class"}" tw="${"bg-red-500"}" tw-2xl="${"bg-red-300"}">test</p>`);';

      const instance = tailpropsPlugin({ framework: "svelte-ssr" });

      const result = instance.transform.bind({
        ...mockedTransformContext,
        meta: { watchMode: false },
      } as any)(source, "test.svelte") as Partial<SourceDescription>;

      expect(result.code).not.toContain("attr_dev");
      expect(result.code).toContain("attr");
    });
  });

  describe("astro integration", () => {
    beforeEach(() => {
      plugin = tailpropsPlugin({ framework: "astro" });
    });

    it("should not touch non-astro files", () => {
      const result = plugin.transform.bind(mockedTransformContext)(
        "console.log('test')",
        "1283901.unknown"
      );

      expect(result).toBeNull();
    });

    it("should support an array of integrations", () => {
      const instance = tailpropsPlugin({
        framework: "astro",
        integrations: [{ name: "react" }],
      });

      const result = instance.transform.bind(mockedTransformContext)(
        "console.log('test')",
        "1283901.jsx"
      ) as Partial<SourceDescription>;

      expect(result.code).toMatch(/console\.log\('test'\);?/);
    });

    it("should parse but not alter files without tailprops", () => {
      const source = 'console.log("test");';

      const result = plugin.transform.bind(mockedTransformContext)(
        source,
        "1283901.astro"
      ) as Partial<SourceDescription>;

      expect(result.code).toBe(source);
    });

    it("should apply template literals transform on Astro files", () => {
      const source =
        'function Test() { return `<html><span${$$addAttribute("test-class", "class")} tw="bg-red-500" tw-hover="bg-red-400">ciao</span></html>`}';

      const instance = tailpropsPlugin({ framework: "astro" });

      const result = instance.transform.bind(mockedTransformContext)(
        source,
        "test.astro"
      ) as Partial<SourceDescription>;

      expect(result.code).toMatchSnapshot();
    });

    it("should support preact integration", () => {
      const instance = tailpropsPlugin({
        framework: "astro",
        integrations: [{ name: "preact" }],
      });

      const result = instance.transform.bind(mockedTransformContext)(
        'function SomeComponent() { return h("div", { className: "test-class", tw: "bg-red-500", "tw-hover": "bg-red-400" }, "ciao"); }',
        "test.jsx"
      ) as Partial<SourceDescription>;

      expect(result.code).toMatchSnapshot();
    });

    it("should support react integration", () => {
      const instance = tailpropsPlugin({
        framework: "astro",
        integrations: [{ name: "react" }],
      });

      const result = instance.transform.bind(mockedTransformContext)(
        'function SomeComponent() { return React.createElement("div", { className: "test-class", tw: "bg-red-500", "tw-hover": "bg-red-400" }, "ciao"); }',
        "test.jsx"
      ) as Partial<SourceDescription>;

      expect(result.code).toMatchSnapshot();
    });
  });

  describe("react integration", () => {
    beforeEach(() => {
      plugin = tailpropsPlugin({ framework: "react" });
    });

    it("should parse but not alter files without tailprops", () => {
      const source = 'console.log("test");';

      const result = plugin.transform.bind(mockedTransformContext)(
        source,
        "1283901.jsx"
      ) as Partial<SourceDescription>;

      expect(result.code).toBe(source);
    });

    it("should ignore non-react files", () => {
      const source = 'console.log("test");';

      const instance = tailpropsPlugin({ framework: "react" });

      const result = instance.transform.bind(mockedTransformContext)(
        source,
        "1283901.svelte"
      ) as Partial<SourceDescription>;

      expect(result).toBeNull();
    });

    it("should apply props object strategy on react files", () => {
      const source =
        'function SomeComponent() { return React.createElement("div", { className: "test-class", tw: "bg-red-500", "tw-hover": "bg-red-400" }, "ciao"); }';

      const instance = tailpropsPlugin({ framework: "react" });

      const result = instance.transform.bind(mockedTransformContext)(
        source,
        "test.jsx"
      ) as Partial<SourceDescription>;

      expect(result.code).toMatchSnapshot();
    });
  });

  describe("preact integration", () => {
    beforeEach(() => {
      plugin = tailpropsPlugin({ framework: "preact" });
    });

    it("should parse but not alter files without tailprops", () => {
      const source = 'console.log("test");';

      const result = plugin.transform.bind(mockedTransformContext)(
        source,
        "1283901.jsx"
      ) as Partial<SourceDescription>;

      expect(result.code).toBe(source);
    });

    it("should ignore non-preact files", () => {
      const source = 'console.log("test");';

      const instance = tailpropsPlugin({ framework: "preact" });

      const result = instance.transform.bind(mockedTransformContext)(
        source,
        "1283901.svelte"
      ) as Partial<SourceDescription>;

      expect(result).toBeNull();
    });

    it("should apply props object strategy on preact files", () => {
      const source =
        'function SomeComponent() { return h("div", { className: "test-class", tw: "bg-red-500", "tw-hover": "bg-red-400" }, "ciao"); }';

      const instance = tailpropsPlugin({ framework: "preact" });

      const result = instance.transform.bind(mockedTransformContext)(
        source,
        "test.jsx"
      ) as Partial<SourceDescription>;

      expect(result.code).toMatchSnapshot();
    });

    it("should account for passed class attribute", () => {
      const source =
        'function SomeComponent() { return h("div", { hoppity: "test-class", tw: "bg-red-500", "tw-hover": "bg-red-400" }, "ciao"); }';

      const instance = tailpropsPlugin({
        framework: "preact",
        classAttribute: "hoppity",
      });

      const result = instance.transform.bind(mockedTransformContext)(
        source,
        "test.jsx"
      ) as Partial<SourceDescription>;

      expect(result.code).toMatchSnapshot();
    });
  });
});
