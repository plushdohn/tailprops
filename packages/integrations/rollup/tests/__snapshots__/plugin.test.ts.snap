// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rollup plugin astro integration should apply template literals transform on Astro files 1`] = `
"function Test() {
  return \`<html><span  \${$$addAttribute("test-class" + " " + "bg-red-500" + " " + "hover:bg-red-400", "class")}>ciao</span></html>\`;
}"
`;

exports[`rollup plugin astro integration should support preact integration 1`] = `
"function SomeComponent() {
  return h("div", {
    className: "test-class" + " " + "bg-red-500" + " " + "hover:bg-red-400"
  }, "ciao");
}"
`;

exports[`rollup plugin astro integration should support react integration 1`] = `
"function SomeComponent() {
  return React.createElement("div", {
    className: "test-class" + " " + "bg-red-500" + " " + "hover:bg-red-400"
  }, "ciao");
}"
`;

exports[`rollup plugin preact integration should account for passed class attribute 1`] = `
"function SomeComponent() {
  return h("div", {
    hoppity: "test-class" + " " + "bg-red-500" + " " + "hover:bg-red-400"
  }, "ciao");
}"
`;

exports[`rollup plugin preact integration should apply props object strategy on preact files 1`] = `
"function SomeComponent() {
  return h("div", {
    className: "test-class" + " " + "bg-red-500" + " " + "hover:bg-red-400"
  }, "ciao");
}"
`;

exports[`rollup plugin react integration should apply props object strategy on react files 1`] = `
"function SomeComponent() {
  return React.createElement("div", {
    className: "test-class" + " " + "bg-red-500" + " " + "hover:bg-red-400"
  }, "ciao");
}"
`;

exports[`rollup plugin svelte integration should apply template literals transform on SSR files 1`] = `"create_ssr_component(\`<p   \${add_attribute("class", "some-class" + " " + "bg-red-500" + " " + "2xl:bg-red-300", 0)}>test</p>\`);"`;
